### DISCUSSION TITLE
Postgresql INET type as a primary key in relationships with psycopg3

### DISCUSSION THREAD

**Discussion by jakubman1** [sqlalchemy/sqlalchemy#10965] [Usage Questions] [answered]
Hello,

I am updating my library to use psycopg3 instead of psycopg2, and it seems that psycopg3 converts INET types to `ipaddress.IPvXInterface` python objects. That should be fine, but it looks like SQLAlchemy can not handle a mix of IPv4 and IPv6 addresses in primary keys since IPv4Interface can not be compared to IPv6Interface, while INET supports both IPv4 and IPv6.
The error only happens on state updates due to SQLAlchemy sorting the instances in a session when using an INET as a primary key in a relationship.

A minimal example for reproducing my issue:

```py
import ipaddress

import sqlalchemy
from sqlalchemy import create_engine, Column
from sqlalchemy.dialects.postgresql import INET
from sqlalchemy.orm import relationship, sessionmaker, declarative_base

Base = declarative_base()


class Net(Base):
    __tablename__ = "net"

    id = sqlalchemy.Column(
        sqlalchemy.Integer,
        sqlalchemy.ForeignKey("logs.id", ondelete="CASCADE"),
        primary_key=True,
    )

    ip = sqlalchemy.Column(INET, primary_key=True)


class Log(Base):
    __tablename__ = 'logs'

    id = Column(sqlalchemy.Integer, primary_key=True, autoincrement=True)
    nets = relationship('Net', cascade="all, delete-orphan")


engine = create_engine('postgresql+psycopg://test:test@localhost:5432/test')

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

Session = sessionmaker(bind=engine)
session = Session()

log = Log(id=1)

# Mix of IPv4 and IPv6 addresses
log.nets.append(Net(
    ip=ipaddress.ip_address('127.0.0.1'),
))
log.nets.append(Net(
    ip=ipaddress.ip_address('abcd::1234'),
))

session.add(log)
session.commit()

print(log.nets)  # Force-load the relationship

session.delete(log)  # Perform a state update
session.commit()  # Crashes on first commit/flush
```

The error is `sqlalchemy.exc.InvalidRequestError: Could not sort objects by primary key; primary key values must be sortable in Python (was: 127.0.0.1 and abcd::1234 are not of the same version)`.

The same code used to work with psycopg2 since it converts INET to strings.

I am not sure if it is a bug in SQLAlchemy or if I am doing something wrong.

Is there a workaround for this? I would like to avoid modifying the database schema, if possible.

I am using Python 3.8, SQLAlchemy 2.0.25 and psycopg 3.1.18

Thanks a lot in advance!

**Comment by zzzeek** [comment, marked as answer]
hi -

if the INET datatype is sortable in PostgreSQL, then psycopg3 should be retaining that behavior in its choice of datatype, so this would be a bug in psycopg3 cc @dvarrazzo 

if you want to alter the kind of Python object handled by the datatype you could try using a [TypeDecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#typedecorator-recipes) to return some other kind of object.

**Reply by dvarrazzo** [reply]
It seems a case too marginal to introduce a wrapper object to stdlib objects just to make ipv4 and ipv6 objects sortable across version (which would be probably an arbitrary ordering).

In psycopg2 the network types were opt-in. In psycopg 3 it's possible to opt out from them by registering the text loader as their loader:

```python
>>> import psycopg
>>> conn = psycopg.connect("", autocommit=True)
>>> conn.execute("select '192.168.0.1'::inet").fetchone()[0]
IPv4Address('192.168.0.1')

>>> conn.adapters.register_loader("inet", psycopg.types.string.TextLoader)
>>> conn.execute("select '192.168.0.1'::inet").fetchone()[0]
'192.168.0.1'
```

I don't know if SA offers support to configure adapters in psycopg 3 or if the OP wants to make this configuration at psycopg level in their code.

**Reply by zzzeek** [reply]
how does PostgreSQL apply sorting and indexing to INET address that can be either ipv4 or ipv6 ?

**Reply by CaselIT** [reply]
> how does PostgreSQL apply sorting and indexing to INET address that can be either ipv4 or ipv6 ?

pg can sort, but it's unclear how it does that. It's not by text and it's not by converting ip4 to ip6. Almost seems like any ip4 are before any ip6

> I don't know if SA offers support to configure adapters in psycopg 3 or if the OP wants to make this configuration at psycopg level in their code.

global adapters will work. It's only currently possible to customize connection adapters using an event, like https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.connect

**Reply by dvarrazzo** [reply]
Postgres sort function is [this](https://github.com/postgres/postgres/blob/1ae5ace7558ea949d2f94af2fd5eb145d5558659/src/backend/utils/adt/network.c#L393-L422), it seems dependent on internal details, I haven't drilled into it.

**Reply by CaselIT** [reply]
Skimming it, that seem like v4 always before v6 like observed by my quick test, since this is what is done in case the versions are different:

`return ip_family(a1) - ip_family(a2);`

Maybe something similar could be proposed to python too?

**Reply by jakubman1** [reply]
I ended up using the TextLoader, thank you for your help.
